app/src/main/java/com/android/voyageur/ui/gallery/PermissionButtonForGallery.kt

```diff
@@ -0,0 +1,144 @@
+            Toast.makeText(
+                    context, "Permission denied. Unable to select photo.", Toast.LENGTH_SHORT)
+                .show()
+          }
+        }
+      }
+  Button(
+      onClick = {
+        when {
+          checkFullPermission(context) || checkLimitedPermission(context) -> {
+            // If permission is granted, launch the gallery
+            galleryLauncher.launch("image/*")
+          }
+          ActivityCompat.shouldShowRequestPermissionRationale(
+              context.findActivity(), permissionVersion) -> {
+            // Show rationale
+            showRationaleDialog = true
+          }
+          else -> {
+            permissionLauncher.launch(permissionVersion)
+          }
+        }
+      },
+      modifier = modifier) {
+        // Shows message corresponding to the screen
+        Text(messageToShow)
+        // Show rationale dialog if needed
+        if (showRationaleDialog) {
+          AlertDialog(
+              onDismissRequest = { showRationaleDialog = false },
+              title = { Text("Permission Required") },
+              text = { Text(dialogMessage) },
+              confirmButton = {
+                TextButton(
+                    onClick = {
+                      showRationaleDialog = false
+                      permissionLauncher.launch(permissionVersion)
+                    }) {
+                      Text("Allow")
+                    }
+              },
+              dismissButton = {
+                TextButton(
+                    onClick = {
+                      showRationaleDialog = false
+                      Toast.makeText(
+                              context,
+                              "Permission denied. Unable to select photo.",
+                              Toast.LENGTH_SHORT)
+                          .show()
+                    }) {
+                      Text("Cancel")
+                    }
+              })
+        }
+      }
+}
+
+fun checkFullPermission(context: Context): Boolean {
+  return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+    ContextCompat.checkSelfPermission(context, READ_MEDIA_IMAGES) == PERMISSION_GRANTED
+  } else {
+    ContextCompat.checkSelfPermission(context, READ_EXTERNAL_STORAGE) == PERMISSION_GRANTED
+  }
+}
+
+fun checkLimitedPermission(context: Context): Boolean {
+  return (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE &&
+      ContextCompat.checkSelfPermission(context, READ_MEDIA_VISUAL_USER_SELECTED) ==
+          PERMISSION_GRANTED)
+}
+/**
+ * Return Component Activity associated with Context to show dialog. Unwraps context on which it is
+ * called upon.
+ * *
+ */
+fun Context.findActivity(): ComponentActivity {
+  var context = this
+  while (context is ContextWrapper) {
+    if (context is ComponentActivity) {
+      return context
+    }
+    context = context.baseContext
+  }
+  throw IllegalStateException("No Activity found")
+}
```

Comment:
Instead of throwing an exception in the findActivity() function, consider returning null when no activity is found. Then, perform a null check wherever this function is called to handle the scenario gracefully and prevent potential crashes.

Reply:
I agree, i think it's better to avoid throwing the exception and cause the app to crash, I'll modify this. Thanks!